src/util/string.h

```diff
@@ -53,23 +46,45 @@ struct ConstevalFormatString {
                 continue;
             }
 
-            unsigned maybe_num{0};
-            while ('0' <= *it && *it <= '9') {
-                maybe_num *= 10;
-                maybe_num += *it - '0';
-                ++it;
+            auto add_arg = [&] {
+                unsigned maybe_num{0};
+                while ('0' <= *it && *it <= '9') {
+                    maybe_num *= 10;
+                    maybe_num += *it - '0';
+                    ++it;
```

Comment:
we still have unbounded increment without checking the end (I though we've fixed this already, maybe it got stuck in the comments...)

Reply:
re: https://github.com/bitcoin/bitcoin/pull/31174#discussion_r1820938133

> we still have unbounded increment without checking the end (I though we've fixed this already, maybe it got stuck in the comments...)

Thanks. I fixed this by just switching the string type to `const char*` instead of `string_view` since tinyformat already assumes the string is null terminated.

I think it would be possible to write a clean version of this code that used `string_view`, but it would have to be structured differently to avoid the need for `it < str.end()` checks everywhere.

Reply:
Found the PR thanks to @hodlinator, this is how I handled that case: https://github.com/bitcoin/bitcoin/pull/30999/files#diff-71badc1cc71ba46244f7841a088251bb294265f4fe9e662c0ad6b15be540eee4R60

We could a test case for the incomplete trailing number (or any other test that seems relevant here): https://github.com/bitcoin/bitcoin/pull/30999/files#diff-718d0d85269ec81595bd9f9181eea3a74b20244b07f14c546e3e07520b2b5f82R81

Reply:
Isn't that done here?
https://github.com/bitcoin/bitcoin/blob/ad37073f2e6ab1f39a59109692f84cc85809f53e/src/test/util_string_tests.cpp#L90

Reply:
It is indeed now, thanks for checking.